WIP: TST: Add xfailing tests for rewrite of numpy.pad

lagru opened the pull request:
Split from #11358.

These tests describe not yet implemented behavior for the pad function:
- pad should raise a explicit error if the "mode" argument is not
  supported.
- pad should preserve the memory-layout (C/F order) of its input when
  padding.

The test coverage is actually at 100%. Coverage.py says that [line 1292](https://codecov.io/gh/numpy/numpy/src/master/numpy/lib/arraypad.py#L1289) is not covered which is actually covered by [this test](https://github.com/numpy/numpy/blob/e26c2990c4828d6f7f2f588d75cd01eecafd53f3/numpy/lib/tests/test_arraypad.py#L792). One can check this with a debugger or by inserting the line `print(pad_before > 0 or pad_after > 0)` before the `continue` statement which is printed and actually corrects coverage.py's output. I believe this may be a bug with coverage.py or am I not noticing something obvious?

Coverage for the rewrite in #11358 is nearly at 100 % too but I still want to cover a few only partially reached lines. Therefore this is still marked as WIP.

lagru committed:
TST: Add xfailing tests for rewrite of numpy.pad

These tests describe not yet implemented behavior for the pad function:
- pad should raise a explicit error if the "mode" argument is not
  supported.
- pad should preserve the memory-layout (C/F order) of its input when
  padding.

seberg commented:
@lagru, I was just going to try and start work through your changes (maybe a bit now, hopefully I find enough time in the next few days). I have never really worked with these xfail test, and such. We can do this if you like, I do not care much either way.

Since it seems to me like this is a complete rewrite, so I doubt that splitting it up to inconsistent chunks will really help reviewing much. If the commits are in meaningful chunks, maybe I will look through them individually.

What is important to me, is that we do not accidentally change behaviour, so if you know of any other possible changes, please say so. It seems the pad tests are not so bad, but I am not sure they are very comprehensive either, so I think I might want some more in the longer run (although this is python code, so the crazier stuff is not necessary).

lagru commented:
@seberg Thanks in advance.

> I have never really worked with these xfail test

Well I had to mark these as [xfailing](https://docs.pytest.org/en/latest/skipping.html) otherwise the CLI wouldn't pass for master if this is merged before #11358.

> What is important to me, is that we do not accidentally change behaviour, so if you know of any other possible changes, please say so. 

The intended changes are covered by these new unit tests and some already marked xfail tests for bugs (e.g. #11216). However I'll add a few tests for the new mode `empty` which is only accidentally covered right now. Other than that I can't think of any other intended changes to pad's behavior.

It may be useful to add more comments on some tests to make their intention more obvious. And maybe reorganize / refactor some (e.g. we have 3 test classes named `TestValueError`). Let me know what you think. ;)

> If the commits are in meaningful chunks, maybe I will look through them individually.

In the beginning yes but not towards the end due to me iterating and addressing reviews. I think you'll only have more work (outdated changes) if you try to follow the commit history.

seberg commented:
I realize what xfail is, but I don't have experience with it helping with PR workflow :).

One thing that I think we should add is a few tests for high dimensions, even if I doubt there is a likely issue with it.

seberg commented:
Those three classes are funny, if they have such a silly name, why not just keep it one class? But meh...

seberg commented on numpy/lib/tests/test_arraypad.py:
```diff
@@ -1260,3 +1260,38 @@ def test_check_wrong_pad_amount(self):
+
+
+@pytest.mark.xfail(reason="not yet implemented, see gh-11358")
+def test_order():
```
I am good with putting this in as is (leaving new tests for later), if you like. One thing I would like here is a test for a case that is not contiguous (3d or 2d slices). And also a test for a 1D array, because it is both F and C-order.

eric-wieser replied:
That test would consist of checking that `np.argsort(padded.strides) == np.argsort(orig.strides)`, for cases where the strides are not pre-sorted.

seberg replied:
I doubt we will add that in the first iteration, nor do I think it is a serious bug to return C-order, so I am not sure we need an xfail test for it (which would stay xfail indefinitely).

seberg commented on numpy/lib/tests/test_arraypad.py:
```diff
@@ -1260,3 +1260,38 @@ def test_check_wrong_pad_amount(self):
+
+@pytest.mark.xfail(reason="not yet implemented, see gh-11358")
+def test_unsupported_mode():
+    """Test error message for unsupported pad modes."""
```
Oh, one thing, unless this changed with pytest, please avoid docstrings and use comments instead. At least nose would replace test function with the doc string, which tends to be a bit annoying.

eric-wieser replied:
pytest does not do this by default, so I think we should leave it as is.

eric-wieser commented:
I'm a fan of committing `xfail`ing tests to identify bugs that need fixing at some point, but I'm not sure there's any value to them for features that aren't yet implemented.

If you can frame the `reason=...` as a bug, then I'd be happy to merge these before your changes - but if they're just testing an upcoming feature, I don't think doing them separately adds any value.

I think `test_unsupported_mode` fits the category of testing for a bug, but `test_order` doesn't make much sense here.

lagru commented:
> but if they're just testing an upcoming feature, I don't think doing them separately adds any value, and I'd rather merge them as part of the same PR that adds the feature.

I see the first as testing a maintenance improvement giving the user a clearer error message. The second one definitely covers a new feature. But one could view preservation of order as the expected behavior for `np.pad` which is just hard to do in the old implementation. 

I got the impulse for this PR from @mattip 's https://github.com/numpy/numpy/pull/11358#issuecomment-446888824. I'm fine with this going either way. But maybe its more productive to just keep it this way.

Anyway I plan on doing some [mutation testing](https://mutmut.readthedocs.io/en/latest/) on the old implementation. If I find any behavior or gaps worth pinning down I'll append these new tests here. That should prepare the already extensive test suite even more for changing the implementation.

mattip commented:
My intent was to enable faster progress toward merging #11358, but I did not realize there are behavioural changes (new keyword). So sorry to cause you all *more* work, the intent was the opposite.

seberg commented:
Hehe, well, frankly, nothing here is so off the rails that it matters much. But it also seems these tests are still on the other PR as well, so we can keep them there, I guess. I am good with being pragmatic here and do whatever @lagru likes.

lagru commented:
> But it also seems these tests are still on the other PR as well, so we can keep them there,

True. Might be better to keep them there rather than introduce tests for features which might change in the future. I'll keep your reviews here in mind and will address them over at #11358.

I'll raise another PR when I have something that improves the tests and coverage of pad's old implementation.
